crates/winch/src/isa/aarch64/abi.rs

```diff
@@ -0,0 +1,215 @@
+use super::regs;
+use crate::abi::{ABIArg, ABIResult, ABISig, ABI};
+use crate::isa::reg::Reg;
+use smallvec::SmallVec;
+use wasmtime_environ::{WasmFuncType, WasmType};
+
+#[derive(Default)]
+pub(crate) struct Aarch64ABI;
+
+/// Helper environment to track argument-register
+/// assignment in aarch64
+///
+/// The first element tracks the general purpose register index, capped at 7 (x0-x7)
+/// The second element tracks the floating point register index, capped at 7 (v0-v7)
+// Follows
+// https://github.com/ARM-software/abi-aa/blob/2021Q1/aapcs64/aapcs64.rst#64parameter-passing
+#[derive(Default)]
+struct RegIndexEnv(u8, u8);
+
+impl RegIndexEnv {
+    fn next_xreg(&mut self) -> Option<u8> {
+        if self.0 < 8 {
+            return Some(Self::increment(&mut self.0));
+        }
+
+        None
+    }
+
+    fn next_vreg(&mut self) -> Option<u8> {
+        if self.1 < 8 {
+            return Some(Self::increment(&mut self.1));
+        }
+
+        None
+    }
+
+    fn increment(index: &mut u8) -> u8 {
+        let current = *index;
+        *index += 1;
+        current
+    }
+}
+
+impl ABI for Aarch64ABI {
+    // TODO change to 16 once SIMD is supported
+    fn stack_align(&self) -> u8 {
+        8
+    }
+
+    fn arg_base_offset(&self) -> u8 {
+        16
+    }
+
+    fn word_bits() -> u32 {
+        64
+    }
+
+    fn sig(&self, wasm_sig: &WasmFuncType) -> ABISig {
+        let mut stack_offset = 0;
+        let mut index_env = RegIndexEnv::default();
+
+        let params: SmallVec<[ABIArg; 6]> = wasm_sig
+            .params()
+            .iter()
+            .map(|arg| Self::to_abi_arg(arg, &mut stack_offset, &mut index_env))
+            .collect();
+
+        let ty = wasm_sig.returns().get(0).map(|e| e.clone());
```

Comment:
We should assert here that there is only one result type, right? (I think that wasmparser probably accepts multi-value Wasm by default now)

Doing the fully general thing is fine to defer to later, but we'll need to support it at some point; in Cranelift our ABI uses the arg regs (x0-x7 on aarch64) as eight returns and then has a "return value area" on the stack for more than that...

Reply:
(Ah, reading further below to the x64 version, I see an assert there, so I think it's just missing here)

Reply:
Added, thanks!